src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -3821,46 +3887,71 @@ internal void CloneCleanupAltMetaDataSetArray()
             }
         }
 
-        private class PacketData
+        private sealed class StateSnapshot
         {
-            public byte[] Buffer;
-            public int Read;
-#if DEBUG
-            public string Stack;
-#endif
-        }
+            private sealed class PLPData
+            {
+                public readonly ulong SnapshotLongLen;
+                public readonly ulong SnapshotLongLenLeft;
 
-        private class StateSnapshot
-        {
-            private List<PacketData> _snapshotInBuffs;
-            private int _snapshotInBuffCurrent = 0;
-            private int _snapshotInBytesUsed = 0;
-            private int _snapshotInBytesPacket = 0;
-            private bool _snapshotPendingData = false;
-            private bool _snapshotErrorTokenReceived = false;
-            private bool _snapshotHasOpenResult = false;
-            private bool _snapshotReceivedColumnMetadata = false;
-            private bool _snapshotAttentionReceived;
-            private byte _snapshotMessageStatus;
+                public PLPData(ulong snapshotLongLen, ulong snapshotLongLenLeft)
+                {
+                    SnapshotLongLen = snapshotLongLen;
+                    SnapshotLongLenLeft = snapshotLongLenLeft;
+                }
+            }
 
-            private NullBitmap _snapshotNullBitmapInfo;
-            private ulong _snapshotLongLen;
-            private ulong _snapshotLongLenLeft;
-            private _SqlMetaDataSet _snapshotCleanupMetaData;
-            private _SqlMetaDataSetCollection _snapshotCleanupAltMetaDataSetArray;
+            private sealed partial class PacketData
+            {
+                public byte[] Buffer;
+                public int Read;
+                public PacketData Prev;
+
+                public void SetStack(string value)
+                {
+                    SetStackInternal(value);
+                }
+                partial void SetStackInternal(string value);
 
-            private readonly TdsParserStateObject _stateObj;
+                public void Clear()
+                {
+                    Buffer = null;
+                    Read = 0;
+                    Prev = null;
+                    SetStackInternal(null);
+                }
+            }
 
-            public StateSnapshot(TdsParserStateObject state)
+#if DEBUG
+            private sealed partial class PacketData
             {
-                _snapshotInBuffs = new List<PacketData>();
-                _stateObj = state;
+                public string Stack;
+
+                partial void SetStackInternal(string value)
+                {
+                    Stack = value;
+                }
             }
 
-#if DEBUG
             private int _rollingPend = 0;
             private int _rollingPendCount = 0;
+#endif
+            private PacketData _snapshotInBuffList;
+            private PacketData _sparePacket;
```

Comment:
What is _sparePacket for? It looks like it is assigned to but never used.

Reply:
Nevermind. I'm just getting blurry-eyed. I see it now.

Reply:
https://github.com/dotnet/SqlClient/blob/dc8a30f2cf3c47701db96c9f4cd28fa1e53289f5/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs#L4016-L4024

It's a cached spare. If you need a packet it's tried first and used and then another is created if needed. If you have a snapshot you always have a packet data so the snapshot carries one with it when it's cached.